Add virtual implementation info to lib info

rgrinberg opened the pull request:
For the first stage, I'll only support virtual libraries and implementations that are defined locally. This is because full support will require refactoring dune-project.

I'm not sure if the way I'm handling requires and ppx rd's for implementations is the best way. Another approach is to take the implementation's info and combine those. But then, we don't really want to resolve the virtual lib's selects again.

## Inehrited fields in Implementations:

* `libraries`
* `ppx_runtime_dependencies`
* `wrapped` (always `true`)

rgrinberg commented:
I also added some simple rules for copying the archives from the virtual library and some error checking for implementations. Since this is all possible with the new lib_info fields.

diml commented:
Could you give an overview of this PR? For instance, I understand that in this feature some fields of implementations are inherited from the virtual library, while some others are not. Could you precise which ones are inherited and which ones are not?

rgrinberg commented:
Okay, this PR started quite a bit smaller in scope but then I extended it with a "couple" of other things. Let me give you a quick summary:

* We need to extend library information to make building implementations possible. This requires exposing the dependency graph + virtual modules. For now, this information will only be obtained for internal libraries (external implementations will be handled later)

* I've added the resolution of the virtual lib for implementations. This is necessary to resolve requires and ppx rd's for implementations.

* I've added some simple code to copy artifacts from the virtual library to the implementation directory.

* I've disabled archives/plugins for virtual libraries (this is straight forward, bunt unrelated to other issues)

* Building stubs for implementations will now use object files from the virtual library.

* I've added some code to do some verifications for implementations. Checks like that all virtual libraries have been provided an implementation.

diml commented:
Ok, thanks for the summary, it makes it a lot easier to read the code!
